crates/autopilot/src/on_settlement_event_updater.rs

```diff
@@ -0,0 +1,325 @@
+        let gas_used = receipt
+            .gas_used
+            .with_context(|| format!("no gas used {hash:?}"))?;
+        let effective_gas_price = receipt
+            .effective_gas_price
+            .with_context(|| format!("no effective gas price {hash:?}"))?;
+
+        let auction_id = self
+            .db
+            .get_auction_id(tx_from, tx_nonce)
+            .await?
+            .context(format!("no auction id for tx {hash:?}"))?;
+        let auction_external_prices =
+            self.db
+                .get_auction_prices(auction_id)
+                .await
+                .context(format!(
+                    "no external prices for auction id {auction_id:?} and tx {hash:?}"
+                ))?;
+        let orders = self
+            .db
+            .orders_for_tx(&hash)
+            .await?
+            .into_iter()
+            .map(Into::into)
+            .collect::<Vec<_>>();
+        let external_prices = ExternalPrices::try_from_auction_prices(
+            self.native_token,
+            auction_external_prices.clone(),
+        )?;
+
+        // surplus and fees calculation
+        let configuration = FeeConfiguration {
+            fee_objective_scaling_factor: self.fee_objective_scaling_factor.clone(),
+        };
+        let settlement = DecodedSettlement::new(&transaction.input.0)?;
+        let surplus = settlement.total_surplus(&external_prices);
+        let fee = settlement.total_fees(&external_prices, &orders, &configuration);
+
+        // reduce external prices in `auction_prices` table to only include used tokens
+        // this is done to reduce the amount of data we store in the database
+        let prices = settlement
+            .trades
+            .iter()
+            .filter_map(|trade| {
+                let buy_token = settlement
+                    .tokens
+                    .get(trade.buy_token_index.as_u64() as usize)
+                    .unwrap(); // TODO can I unwrap here? Is it guaranteed by the settlement contract?
+                let sell_token = settlement
+                    .tokens
+                    .get(trade.sell_token_index.as_u64() as usize)
+                    .unwrap(); // TODO can I unwrap here? Is it guaranteed by the settlement contract?
+                let buy_token_price = auction_external_prices.get(buy_token);
+                let sell_token_price = auction_external_prices.get(sell_token);
+                match (buy_token_price, sell_token_price) {
+                    (Some(buy_token_price), Some(sell_token_price)) => Some(vec![
+                        (*buy_token, *buy_token_price),
+                        (*sell_token, *sell_token_price),
+                    ]),
+                    _ => {
+                        tracing::error!("settlement used token that was not in auction");
+                        None
+                    }
+                }
+            })
+            .flatten()
```

Comment:
nit: You could skip this call if you used a `.flat_map()` (returning `vec![]` instead of `None`).